activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,160 @@
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(Key)
+
+        self.new(key)
+      end
+
+      def initialize(key = nil)
+        @cache_key = +""
+        @cache_version = +""
+        @empty_key = true
+        self << key
+      end
+
+      def update(key)
+        if key.respond_to?(:cache_key)
+          store_key(key.cache_key.to_s)
+          store_version(key)
+          return
+        end
+
+        case key
+        when Array
+          key.each do |k|
+            self << k
+          end
+        when Hash
+          key = key.to_a
+          store_version(key)
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+
+          key.each do |k|
+            store_key(k)
+          end
+        else
+          store_version(key)
+          store_key(key)
+        end
+      end
+      alias :<< :update
+
+      private
+        def store_key(key)
+          @cache_key << "/" unless @empty_key
+          @empty_version =
+          @cache_key << key.to_param.to_s
+          @empty_key = false
+        end
+
+        def store_version(key)
+          case
+          when key.respond_to?(:cache_version)
+            @cache_version << "/" unless @cache_version.empty?
+            @cache_version << key.cache_version.to_param.to_s
+          when key.is_a?(Array)
+            key.each do |k|
+              store_version(k)
+            end
+          when key.respond_to?(:to_a)
+            store_version(key.to_a)
```

Comment:
On my initial read at least, I felt like it would make sense to combine this with the above logic or duplicate the above logic so that you're not adding an extra layer of recursive logic.

Reply:
I don't totally follow here. This method represents the existing logic of `expanded_version` while the main body of `<<` represents the existing logic of `expanded_key`. 

We cannot get rid of this `to_a` behavior for the case of a hash containing another object that can respond to `to_a` such as another hash.

I'll add this as a test case to ensure this logic is preserved

```
    klass = Class.new { def cache_version; "foo"; end }
    key = ActiveSupport::Cache::Key.new({hello: {world: klass.new}})
    assert_equal "foo", key.cache_version
```